fhirpath/src/test/java/au/csiro/pathling/fhirpath/parser/FhirpathTest.java

```diff
@@ -0,0 +1,115 @@
+import au.csiro.pathling.test.SpringBootUnitTest;
+import au.csiro.pathling.test.assertions.FhirPathAssertion;
+import au.csiro.pathling.test.builders.ParserContextBuilder;
+import ca.uhn.fhir.context.FhirContext;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.apache.spark.sql.SparkSession;
+import org.hl7.fhir.r4.model.Condition;
+import org.hl7.fhir.r4.model.Enumerations.ResourceType;
+import org.hl7.fhir.r4.model.Identifier;
+import org.hl7.fhir.r4.model.Reference;
+import org.hl7.fhir.r4.model.Resource;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+
+@SpringBootUnitTest
+public class FhirpathTest {
+
+  @Autowired
+  protected SparkSession spark;
+
+  @Autowired
+  FhirContext fhirContext;
+
+  @Autowired
+  TerminologyService terminologyService;
+
+  @Autowired
+  FhirEncoders fhirEncoders;
+
+  @Autowired
+  TerminologyServiceFactory terminologyServiceFactory;
+
+  @MockBean
+  DataSource dataSource;
+
+
+  @SuppressWarnings("SameParameterValue")
+  @Nonnull
+  protected FhirPathAssertion assertThatResultOf(@Nonnull final ResourceType resourceType,
+      @Nonnull final String expression) {
+    final ResourcePath subjectResource = ResourcePath
+        .build(fhirContext, dataSource, resourceType, resourceType.toCode(), true);
+
+    final ParserContext parserContext = new ParserContextBuilder(spark, fhirContext)
+        .terminologyClientFactory(terminologyServiceFactory)
+        .database(dataSource)
+        .inputContext(subjectResource)
+        .build();
+    final Parser resourceParser = new Parser(parserContext);
+    return assertThat(resourceParser.parse(expression));
+  }
+
+
+  void withResources(@Nonnull final Resource resources) {
+
+    // group resources by type
+    // and then encode them into a dataset and setup the mock datasorce
+
+    Stream.of(resources).collect(Collectors.groupingBy(Resource::getResourceType))
+        .forEach((resourceType, resourcesOfType) -> {
+          final ResourceType resourceTypeEnum = ResourceType.fromCode(resourceType.name());
+          final Dataset<Row> dataset = spark.createDataset(resourcesOfType,
+                  fhirEncoders.of(resourceTypeEnum.toCode()))
+              .toDF();
+          when(dataSource.read(resourceTypeEnum)).thenReturn(dataset.cache());
+        });
+  }
+
+
+  @Test
+  void testTraversalIntoReferenceIdentifier() {
+    withResources(
+        new Condition()
+            .setSubject(new Reference().setIdentifier(new Identifier().setValue("value")))
+            .setId("001")
+    );
+    assertThatResultOf(ResourceType.CONDITION, "subject.identifier.value")
+        .selectResult()
+        .hasRows(RowFactory.create("001", "value"));
+  }
```

Comment:
Ok, I can see that you have actioned the previous comment here. 

I like the idea, but should we also bring the other tests over to this new scheme?

Reply:
Possibly, but I think that should be a separate task and maybe somehow related to the testing of columnar representation. We my though keep using the new test for testing new features/changes in the parser.
Also, we should be able to test that in the existing ParserTest even though we do not have any instances of  non- empty Reference.identifiers  - but that should not matter in this case.

Reply:
Ok. I think this might come under general refactoring of the testing strategy within #1775.